### PR TITLE
Ports Hippie's progress bar animations

### PR DESCRIPTION
**Pull request by actioninja**
## About The Pull Request

Adds some nice animations for progress bars, ported from Hippiecode.
https://youtu.be/aiAdX2g3HUI
Also makes the final stage of progress bars gold instead of green to make it more obvious when it's finished.
Progress bars blink red while fading out if the action is interrupted for some reason.

## Why It's Good For The Game

Looks good.

## Changelog
:cl:
tweak: Some new progress bar animations ported from Hippiestation
/:cl:

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Only dislike is that the current lack of animation lets you know exactly when you're done doing something by immediately disappearing the frame you're done, while these fade out slowly instead. Maybe have the bar snap to a gold color or something the frame they finish so they can keep this cue while still having a fade out animation?

**Comment by MrPerson** [comment]
Green would normally be more associated with completion, but honestly either would work.

**Comment by actioninja** [comment]
> Only dislike is that the current lack of animation lets you know exactly when you're done doing something by immediately disappearing the frame you're done, while these fade out slowly instead. Maybe have the bar snap to a gold color or something the frame they finish so they can keep this cue while still having a fade out animation?

this is beyond nitpicking but ok.

**Comment by MrDoomBringer** [comment]
@actioninja 

>beyond nitpicking

i disagree. His points were valid and this PR is such a minor change in and of itself that small, nitpicky suggestions affect this PR by a huge amount

**Comment by PKPenguin321** [comment]
Really like how it looks with my suggestion, thanks for implementing it!

Question: How does the animation handle it if you fail the task before it finishes? Does it just fade out early?

**Comment by actioninja** [comment]
> Really like how it looks with my suggestion, thanks for implementing it!
> 
> Question: How does the animation handle it if you fail the task before it finishes? Does it just fade out early?

I couldn't remember off the top of my head, so I had to go test. It disappears immediately which seems a little odd. I might tweak it so that it freezes the current progress and fades away or something like that.

**Comment by PKPenguin321** [comment]
If you decide to make it fade out, maybe you could also make it flash red the instant it fails as a counterpart to the gold color on success. Would be neat

File: code/datums/progressbar.dm [hunk 1]
```diff
@@ -59,9 +63,13 @@
 	if(!bars.len)
 		LAZYREMOVE(user.progressbars, bar.loc)
 
-	if (client)
-		client.images -= bar
-	qdel(bar)
+	animate(bar, alpha = 0, time = 5)
+	addtimer(CALLBACK(src, /datum/progressbar/proc/remove_from_client), 5)
+	QDEL_IN(bar, 5)
```
Comment (Fox-McCloud) [inline code/datums/progressbar.dm]: Make sure to test that progress bars will still garbage collect after doing this change.

File: code/datums/progressbar.dm [hunk 2]
```diff
@@ -59,9 +63,13 @@
...
 	. = ..()
 
-#undef PROGRESSBAR_HEIGHT
\ No newline at end of file
+/datum/progressbar/proc/remove_from_client()
+	if(client)
+		client.images -= bar
```
Comment (Fox-McCloud) [inline code/datums/progressbar.dm]: should probably null out the client too, just in case.

File: code/datums/progressbar.dm [hunk 3]
```diff
@@ -59,9 +63,13 @@
 	if(!bars.len)
 		LAZYREMOVE(user.progressbars, bar.loc)
 
-	if (client)
-		client.images -= bar
-	qdel(bar)
+	animate(bar, alpha = 0, time = 5)
+	addtimer(CALLBACK(src, /datum/progressbar/proc/remove_from_client), 5)
```
Comment (Fox-McCloud) [inline code/datums/progressbar.dm]: This needs `TIMER_CLIENT_TIME`

when you use `animate`, you need more precise timers.

**Comment by ATH1909** [comment]
> If you decide to make it fade out, maybe you could also make it flash red the instant it fails as a counterpart to the gold color on success. Would be neat

But don't we already use red in the animation?

**Comment by knacker48** [comment]
> > If you decide to make it fade out, maybe you could also make it flash red the instant it fails as a counterpart to the gold color on success. Would be neat
> 
> But don't we already use red in the animation? Shouldn't we use a different failure color instead?

You could always use a darker shade or red if that's actually a problem

**Comment by PKPenguin321** [comment]
Note that I said *blinking* red. The frame it fails, it could turn off, then blink on, then blink off, etc. (at a fairly high speed)

File: code/datums/progressbar.dm [hunk 4]
```diff
@@ -59,9 +63,14 @@
 	if(!bars.len)
 		LAZYREMOVE(user.progressbars, bar.loc)
 
-	if (client)
-		client.images -= bar
-	qdel(bar)
+	animate(bar, alpha = 0, time = 5)
+	addtimer(CALLBACK(src, /datum/progressbar/proc/remove_from_client), 5, TIMER_CLIENT_TIME)
```
Comment (Fox-McCloud) [inline code/datums/progressbar.dm]: shouldn't this be `addtimer(CALLBACK(src, .proc/remove_from_client), 5, TIMER_CLIENT_TIME)`?

**Comment by actioninja** [comment]
Upon further testing I'm an idiot who tests things on the wrong branch. It actually just fades out upon failure. I'm working on the blinking red idea right now though.

File: code/datums/progressbar.dm [hunk 5]
```diff
@@ -59,9 +68,15 @@
 	if(!bars.len)
 		LAZYREMOVE(user.progressbars, bar.loc)
 
-	if (client)
-		client.images -= bar
-	qdel(bar)
+	animate(bar, alpha = 0, time = PROGRESSBAR_ANIMATION_TIME)
+	addtimer(CALLBACK(src, .proc/remove_from_client), PROGRESSBAR_ANIMATION_TIME, TIMER_CLIENT_TIME)
+	QDEL_IN(bar, PROGRESSBAR_ANIMATION_TIME + 1) //for garbage collection safety
```
Comment (optimumtact) [inline code/datums/progressbar.dm]: this really should be longer

**Pull request merged**